### PR TITLE
Avoid reporting empty version_info for components that just started

### PR DESCRIPTION
**Pull request by AndersonQ**
## What does this PR do?

The state diagnostics hook now re-fetches once the state when necessary, improving the accuracy of component status reporting.

## Why is it important?

Given a design choice, the coordinator's reported state might be stale, leading to healthy components lacking version_info immediately after they start.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- ~~[ ] I have made corresponding changes to the documentation~~
- ~~[ ] I have made corresponding change to the default configuration files~~
- [x] I have added tests that prove my fix is effective or that my feature works
- ~~[ ] I have added an entry in `./changelog/fragments` using th~~e [changelog tool](https://github.com/elastic/elastic-agent#changelog)
- ~~[ ] I have added an integration test or an E2E test~~

## How to test this PR locally

Run the [`TestDiagnosticState`](https://github.com/AndersonQ/elastic-agent/blob/db1be7588b892df7d2ca902d5edd49727ae766d8/internal/pkg/agent/application/coordinator/diagnostics_test.go#L413) test

## Related issues

- Closes https://github.com/elastic/elastic-agent/issues/4215

## Questions to ask yourself

- How are we going to support this in production?
- How are we going to measure its adoption?
- How are we going to debug this?
- What are the metrics I should take care of?
- ...

### CODE REVIEW

**Review by blakerouse — CHANGES_REQUESTED** [review]
I don't fully understand why this is being done? Do we really need it to have that guarantee? I don't feel like we do.

I don't like the idea that we are adjust core code to fix a condition that the test should just probably handle.

Why not have the test fetch the state twice, instead of doing it internally in agent?

**Comment by AndersonQ** [comment]
> Why not have the test fetch the state twice, instead of doing it internally in agent?

my reasoning was that we choose to allow that to happens, so we might as well try to prevent it to happen in the diagnostics. As it's just the diagnostics hook code, I believe it's localised enough to do not lead to unexpected behaviour.

Also it's been happening on the tests, making it not that unlikely as expected

**Comment by blakerouse** [comment]
I don't really see how this is really fixing anything, as it just calls the same function again in that one case. So its just reducing the chance that it could happen, but it still will happen.

**Comment by AndersonQ** [comment]
This is the build info that should come during the check-in, the first one. Therefore most likely the 1st check-in happens, the coordinator internally have it all updated but because of the broadcaster when the diagnostics hook runs, the received state is stale. However a second call should return the right state because by the time the call happens again 1) the stale value has been read by the 1st call 2) the new value is on the broadcaster. Thus a 2nd call should be enough.

**Comment by blakerouse** [comment]
@AndersonQ how are your confirming that the new value is present?

**Comment by AndersonQ** [comment]
> @AndersonQ how are your confirming that the new value is present?

I'm not. If the component is healthy, the check-in happened and the versionInfo was sent. Thus there isn't much reason to believe several retries would be needed. Besides if it were the case the versionInfo isn't be correctly updated 1) the diagnostics should proceed and not get stuck anyway 2) we'd catch it on our tests. Also the diagnostic from the flaky test has all then versionInfo, shoing a retry might be all we need.

**Comment by blakerouse** [comment]
> > @AndersonQ how are your confirming that the new value is present?
> 
> I'm not. If the component is healthy, the check-in happened and the versionInfo was sent. Thus there isn't much reason to believe several retries would be needed. Besides if it were the case the versionInfo isn't be correctly updated 1) the diagnostics should proceed and not get stuck anyway 2) we'd catch it on our tests. Also the diagnostic from the flaky test has all then versionInfo, shoing a retry might be all we need.

It is still unclear how a single retry provides the guarantee that versionInfo is now set. That is what I do not understand, and why I am not okay with this PR at the moment. Currently in my head this change is just reducing the chance slightly, but it is still possible.

**Comment by AndersonQ** [comment]
> It is still unclear how a single retry provides the guarantee that versionInfo is now set. That is what I do not understand, and why I am not okay with this PR at the moment. Currently in my head this change is just reducing the chance slightly, but it is still possible.

The healthy status and the version_info should arrive together, therefore it should be pretty much instantaneous the update. That's why I believe it more than a slight reduction on the chance of getting the an empty version_info. 

Anyway I added more retries with a timeout between them. If all the retries run it's a 250ms + the normal code execution time, what should be a pretty much negligible on the total time to get a diagnostics. Does it looks more robust now?

File: internal/pkg/agent/application/coordinator/coordinator.go [hunk 1]
```diff
@@ -925,13 +924,19 @@ func (c *Coordinator) DiagnosticHooks() diagnostics.Hooks {
...
+				var compStates []StateComponentOutput
+			outerLoop:
+				for i := 0; i < 5; i++ {
+					s, compStates = getState(c)
+					for _, cs := range compStates {
+						if cs.State.State == client.UnitStateHealthy &&
+							cs.State.VersionInfo.BuildHash != "" &&
+							cs.State.VersionInfo.Meta["commit"] != "" {
+							break outerLoop
+						}
```
Comment (blakerouse) [inline internal/pkg/agent/application/coordinator/coordinator.go:937]: This whole block would be much better as its own function either in `getState` or another wrapper around it.

I still am not a fan of this, again your just improving the odds and not solving the true issue. Could we some how guarantee that when status is updated to Healthy that the VersionInfo is set at the same time? Why are they not?

If there is a bad component that never provides the versioninfo it is going to delay this result as well indefinitely by 250 milliseconds.

**Comment by AndersonQ** [comment]
> [...] again your just improving the odds and not solving the true issue

that's the whole idea. The broadcaster was built with the possibility of reporting stale states and the coordinator uses it that way. Changing it all isn't worth it, thus improving the odds seems good enough. 

the broadcaster might report a stale state:
https://github.com/elastic/elastic-agent/blob/729636ac6a4d83286f4bfca3e8e0d2a2b1a46a6f/pkg/utils/broadcaster/broadcaster.go#L194-L199

and how the coordinator uses it:
https://github.com/elastic/elastic-agent/blob/41ee2bb2f8e96ff902d0c88ed19ac5e3734973b9/internal/pkg/agent/application/coordinator/coordinator.go#L372-L387

the new state is sent on `refreshState()`
https://github.com/elastic/elastic-agent/blob/c929f79c7af4ea99a24104f155eb3a6e20910558/internal/pkg/agent/application/coordinator/coordinator_state.go#L127-L132

triggered by the coordinator's runLoop:
https://github.com/elastic/elastic-agent/blob/41ee2bb2f8e96ff902d0c88ed19ac5e3734973b9/internal/pkg/agent/application/coordinator/coordinator.go#L1095-L1099

the broadcaster will update the value its `Get` sends as soon as it receives a new value:
https://github.com/elastic/elastic-agent/blob/729636ac6a4d83286f4bfca3e8e0d2a2b1a46a6f/pkg/utils/broadcaster/broadcaster.go#L542

there is a lot of indirection until the new state finally arrives at the broadcaster to be consumed by the diagnostics hook. That's why I'm not trying to fix it completely, just avoid the problem with a good old retry. 

Our integration tests will say how good the fix is and we can adjust accordingly. That's why I rather have a single retry for now, which would have minimal impact on the time spent collecting a diagnostics and I believe should be enough to get an up-to-date state.

**Comment by blakerouse** [comment]
I still don't like this change, even with the explanation. Honestly I don't believe it is even fixing the problem, just masking it with better odds.

If the component is being updated with healthy which the package_version_test.go is ensuring that it is healthy then it should have that version information.

Another option is to just adjust the test and make the `allHealthy` function to ensure that version information is present before reporting that the Elastic Agent is healthy. https://github.com/elastic/elastic-agent/blob/main/testing/integration/package_version_test.go#L94

You can use the `c.VersionInfo` to verify that the status information has the version information. If its present there then it should defiantly have it in the diagnostics.

**Comment by AndersonQ** [comment]
> Another option is to just adjust the test and make the allHealthy function to ensure that version information is present before reporting that the Elastic Agent is healthy. https://github.com/elastic/elastic-agent/blob/main/testing/integration/package_version_test.go#L94

This does not address the problem at all, it just hides the issue. That's why I'm against that approach.

> Honestly I don't believe it is even fixing the problem, just masking it with better odds.

That's the idea ;)
It isn't supposed to be a perfect fix. The issue does not compromise the normal behaviour of the agent and a complete fix isn't worth it. That's why I chose a palliative fix, improving the odds of it succeeding with minimal overhead and added complexity.

We're seeing this test failing on CI and tweaking the test to ignore the issue isn't a good approach or a good precedent. If we merge this in the current state we'll be able to see if the problem persists or not. If it persists, another approach is needed, if it doesn't, problem solved.
This at least attempt to address the real issue instead of just ignoring it. Besides all the evidence (the failed tests) suggest a retry might be enough.

**Comment by blakerouse** [comment]
> > Another option is to just adjust the test and make the allHealthy function to ensure that version information is present before reporting that the Elastic Agent is healthy. https://github.com/elastic/elastic-agent/blob/main/testing/integration/package_version_test.go#L94
> 
> This does not address the problem at all, it just hides the issue. That's why I'm against that approach.
> 
> > Honestly I don't believe it is even fixing the problem, just masking it with better odds.
> 
> That's the idea ;) It isn't supposed to be a perfect fix. The issue does not compromise the normal behaviour of the agent and a complete fix isn't worth it. That's why I chose a palliative fix, improving the odds of it succeeding with minimal overhead and added complexity.
> 
> We're seeing this test failing on CI and tweaking the test to ignore the issue isn't a good approach or a good precedent. If we merge this in the current state we'll be able to see if the problem persists or not. If it persists, another approach is needed, if it doesn't, problem solved. This at least attempt to address the real issue instead of just ignoring it. Besides all the evidence (the failed tests) suggest a retry might be enough.

Being that we know that its possible for the version information to be delayed (which even based on all the explanation in the PR, still doesn't make sense on why it would be separate from a single Healthy update), and we accepting that delay. My suggestion is to just fix the test to ensure that its always there before performing diagnostics as the test is explicitly testing contents of diagnostic output.

I don't think we actually care in the real world that it is missing on a freshly started running Elastic Agent. That is why I am suggesting to not to make a change inside the core of the Elastic Agent at all. I have not seen an argument that ensures that we need this guarantee, except in this one test case.

**Review by pchila — COMMENTED** [review]
After reading the whole discussion and having a look at the code modification I tend to agree with @blakerouse about fixing this in test code rather than adding retried to the diagnostic hooks.

If we accept that the version_info will eventually be available and dumped in the diagnostics I would prefer to change the asserts on the test to reflect that.

Conversely, if we need some stronger guarantees about the state returned by the broadcaster it's a much bigger change but it's probably not worth it since the chance of diagnostics generated really quickly after components performed their first checkin is rather small and we can always ask for more diagnostics

**Comment by AndersonQ** [comment]
There is something off, but let's get the flaky failures out of the way for now. I think this solution is good enough for now.

**Review by blakerouse — APPROVED** [review]
Thank you for just adjusting the test!

Looks good.

**Pull request merged**